### DISCUSSION TITLE
Call actions within actions, in a typesafe way?

### DISCUSSION THREAD

**Discussion by isaachinman** [pmndrs/zustand#750] [Q&A] [answered]
I've been using zustand very happily for over a year now, but have frequently run into the problem which I will now describe.

It's a very common pattern in my usage to need to call an action from within another action. As a TypeScript user, the main issue is that the (very well-written) typings from `combine` don't really permit this type of usage out of the box.

Here is an example. It's contrived, but demonstrates the issue:

```tsx
import axios from 'axios'
import create, { GetState, SetState } from 'zustand'
import { combine } from 'zustand/middleware'

export type UserStore = {
  data: null | {
    classes?: string[]
    students?: string[]
    userType: 'teacher' | 'student'
  }
}

export const userStoreInitialState = (): UserStore => ({
  data: null,
})

const userStoreActions = (
  set: SetState<UserStore>,
  get: GetState<UserStore>,
) => ({
  fetchClasses: async () => {
    set({
      data: {
        classes: (await axios('/my-backend/user-type')).data,
        userType: 'student',
      },
    })
  },
  fetchStudents: async () => {
    set({
      data: {
        students: (await axios('/my-backend/user-type')).data,
        userType: 'teacher',
      },
    })
  },
  fetchUserType: async () => {
    set({
      data: {
        userType: (await axios('/my-backend/user-type')).data,
      },
    })
  },
  initialize: async () => {
    const {
      fetchUserType,
      fetchClasses,
      fetchStudents,
    } =
      // TODO: Solve problem of calling actions within actions in a typesafe way
      get() as any

    await fetchUserType()

    const userType = get().data?.userType

    if (userType === 'teacher') {
      fetchStudents()
    } else if (userType === 'student') {
      fetchClasses()
    }
  },
})

export const useUserStore = create(
  combine(
    userStoreInitialState(),
    userStoreActions,
  ),
)
```

So, we want to be able to call other actions, like `fetchStudents`, within an `initialise` action, but obviously we'd like to be able to call `fetchStudents` in isolation, as a standalone action, at other points in the application flow.

I'd be very grateful if anyone has any non-hacky suggestions as to how to make this work in a typesafe way. Thanks!

**Comment by dai-shi** [comment]
Yeah, `combine` can't do that now. I thought it's impossible, but let's welcome @devanshj here and hear their idea.

**Reply by devanshj** [reply]
One obvious way is to write types for actions too: https://tsplay.dev/N9pdjw This has only one assertion at the end in `combine` but other than it's fully type-safe. For example if you forget to implement an action or some property for the initial data it won't compile.

But I think what you want to achieve is infer the actions types. This is not possible because of the recursive nature

```ts
// https://tsplay.dev/w6X0Gm

declare const createActions: <T>(f: (get: () => T) => T) => T

createActions(get => ({
  foo: () => {},
  bar: () => {
    get().foo()
  }
}))
```

There's no way to make the above code compile without type annotations.

**Comment by frontshift** [comment, marked as answer]
I only started out yesterday with Zustand, so forgive me if this is not good practice. I would like to use Immer and also plan to always declare my actions outside of the store so that all my actions actions
- don't have to be accessed via the store as in `const increasePopulation = useStore(state => state.increasePopulation)`
- Don't have to be added to the store interface
- Can easily be called from anywhere within my app without using `useStore.getState().myAction` (I don't have to reference my store anywhere except in the actions file)

I created a base store which only has a `set()` function, which basically just acts as a proxy for Immer's `produce()` method.


```typescript
// TODO: Using NamedSet her as the 3rd parameter "name" is not defined on SetState
export const makeImmutableSetter =
  <Store extends object>(set: NamedSet<Store>): Setter<Store> =>
  (fn, name?: string) =>
    set(produce(fn), false, name);
```

I then created a utility function which merges the base store with the actual store that I want to create (one with `devtools` and one without)

```typescript
export const getBaseStore = <T extends object>(set: SetState<T>): BaseStore<T> => ({
  set: makeImmutableSetter(set),
});

export const createVanillaStoreWithDevtoolsFromBase = <T extends object>(
  getStore: (set: SetState<T & BaseStore<T>>, get?: GetState<T & BaseStore<T>>) => T,
  name: string
) =>
  createVanillaStore<T & BaseStore<T>>(
    devtools(
      (set: SetState<T & BaseStore<T>>, get?: GetState<T & BaseStore<T>>) => ({
        ...getBaseStore(set),
        ...getStore(set, get),
      }),
      { anonymousActionType: name }
    )
  );

```

and this is what my store looks like:
```typescript
import createHookFromStore from 'zustand';
import { makeActionType, createVanillaStoreWithDevtoolsFromBase } from '@stores/utils';

export const actionType = makeActionType('appStore');

export interface AppStore {
  shouldShowPreloader: boolean;
}

export const appStore = createVanillaStoreWithDevtoolsFromBase<AppStore>(
  (set, get) => ({
    shouldShowPreloader: false,
  }),
  'appStore'
);

export const useAppStore = createHookFromStore<AppStore>(appStore);
```

I then have a separate actions file for all my actions where I use the custom `set() function of the base store (using Immer) to set the state.

```typescript
import { actionType, appStore, AppStore } from './app.store';

export const setImmutableState = appStore.getState().set;

export const showPreloader = () => {
  setImmutableState((state) => {
    state.shouldShowPreloader = true
  }, actionType('showPreloader'));
};

export const hidePreloader = () => {
  setImmutableState((state) => {
    state.shouldShowPreloader = false
  }, actionType('hidePreloader'));
}; 


export const fetchStuff = async () => {
  showPreloader()
  const response = await fetchStuff()

  hidePreloader()
  doOtherStuff()

  if (response.data) {
    setImmutableState((state) => {
      state.stuff = response.data;
    }, actionType('fetchStuff'));
  }
}
```

It seems to be working ok from my limited experience and everything is nicely decoupled and I can call actions from within actions. But perhaps there are better solutions?

**Reply by isaachinman** [reply]
This is an interesting take on it. @dai-shi Do you see any drawbacks to this approach of using `myStore.getState().set` in decoupled functions instead of using `compose` as intended?

**Reply by dai-shi** [reply]
From the library perspective, there are no drawbacks. Instead of `myStore.getState().set`, you can just use `myStore.setState`. It doesn't seem a self-contained store, which some people like, but zustand is unopinionated about how it's used.

There's a wiki page slightly related to this: https://github.com/pmndrs/zustand/wiki/Practice-with-no-store-actions

**Reply by frontshift** [reply]
I'm only exposing a `set` function to wrap `produce(fn)`, so I don't have to repeat myself all the time.
I'm beginning to think that it might be better to implement this as middleware to eliminate the need for a base store.

 If you don't use Immer (or do other stuff when you update the state), there is not much of a point I guess, as you can just call `store.setState()` from your actions

**Reply by frontshift** [reply]
@dai-shi To be honest, I don't really see the benefits of having a self-contained store. I can really only see the drawbacks
- I have to define the actions on the store interface
- I have to get a reference to the store itself from within actions inside the store
- Out of the box I would have to treat actions as data that can change when accessing in React (`const increasePopulation = useStore(state => state.increasePopulation`) which doesn't make much sense to me. This should only really apply to the data itself.
- To access actions from outside a React component, I would have to create a new wrapper for the action if I want to avoid tight coupling between my modules and the store 

I might not see the full picture yet, but from that perspective I don't really see any benefits.
It all depends on how you organise your stores of course. If you have stuff flying around everywhere and you don't abstract your store access away, that's probably not great.
Ideally you want to be able to export an entire folder that has all the store dependencies
This is what my structure looks like

```
.
└── stores/
    ├── app/
    │   ├── app.store.ts
    │   ├── app.actions.ts
    │   └── app.subs.ts
    └── user/
        ├── user.store.ts
        ├── user.actions.ts
        └── user.subs.ts
```

**Reply by frontshift** [reply]
Also notice that I am not using store slices here but multiple stores. The reason for this is that I couldn't get this working without middleware that I'd have to create.
The suggested approach I saw was to merge the slices like this
```typescript
const useStore = create( (set, get) => ({
    ...createBearSlice(set, get),
    ...createFishSlice(set, get)
}))
```

My intuition is, that you want to namespace your slices like this:

```typescript
const useStore = create( (set, get) => ({
    bear: createBearSlice(set, get),
    fish: createFishSlice(set, get)
}))
```
But that is not quite as simple since the `set` and `get` functions passed to the slice creators, have the interface of the entire store.
So when I  set/get data on the fish slice, I would have to access everything on the `fish` object from the within the slice, which is not really what you want.
What's kinda missing here for me is something like `combineReducers` in redux, where each slice is then namespaced on the main store but the slice object doesn't know about the store structure. 
This is probably possible with middleware that takes the object name of the slice it's operating on.

**Reply by dai-shi** [reply]
> I don't really see the benefits of having a self-contained store.

I don't personally either. But, I guess quite a few people are used to it.

> My intuition is, that you want to namespace your slices

There's some discussion previously, and seems like we have a third-party lib.
See: #178

**Reply by frontshift** [reply]
Thanks that's good to know. Do you know if this will work with v4 as well?  I think I will keep using the multiple stores setup until this middleware has gained some traction and maturity. Looks like it's just been created. But looks good so far.

**Reply by dai-shi** [reply]
> Do you know if this will work with v4 as well?

Not sure which part is your concern, but what is working in v3 in JS will work in v4 too.

**Reply by frontshift** [reply]
Nothing specifically, just saw some mentions of incompatibility between v3 and v4 for certain techniques. I guess that's related to supporting React concurrent mode and may not affect the public api...

**Reply by dai-shi** [reply]
In v4, we drop some old api (which is already deprecated in the latest v3) and add some breaking changes in types especially with middleware.

**Reply by ddanielgal** [reply]
@frontshift your solution is exactly what I'm looking for, but I'm having a hard time implementing it. Can you share a working codesandbox so I can experiment with it?

Has anybody managed to solve this problem with the new v4 API?

**Reply by ddanielgal** [reply]
I made a very minimal example with the basic idea of defining a setter in the store and accessing it with `getState`. [Here is a codesandbox link.](https://codesandbox.io/s/zustand-external-actions-5jhs7l) This is for zustand v4.

Does anybody see any potential bugs or pitfalls with this solution?

```typescript
// store.ts
import { Draft } from "immer";
import create from "zustand";
import { immer } from "zustand/middleware/immer";

type State = {
  count: number;
};

type Setter = {
  set: (setter: (state: Draft<State>) => void) => void;
};

const initialState: State = { count: 0 };

const useCounter = create<State & Setter>()(
  immer((set) => ({ ...initialState, set: (state) => set(state) }))
);

export default useCounter;
```

```typescript
// actions.ts
import useCounter from "./store";

const set = useCounter.getState().set;

export function increment() {
  set((state) => {
    state.count = state.count + 1;
  });
}

export function decrement() {
  set((state) => {
    state.count = state.count - 1;
  });
}
```

**Reply by dai-shi** [reply]
If typescript doesn't complain, it should be good.

If I were to do that, I would a) use `.setState`, and b) use `produce` directly:

```ts
// store.ts
import create from "zustand";

type State = {
  count: number;
};

const initialState: State = { count: 0 };

const useCounter = create(() => initialState);

export default useCounter;
```

```ts
// actions.ts
import produce from "immer";
import useCounter from "./store";

export function increment() {
  useCounter.setState(produce((state) => {
    state.count = state.count + 1;
  }));
}

export function decrement() {
  useCounter.setState(produce((state) => {
    state.count = state.count - 1;
  }));
}
```

See also: https://github.com/pmndrs/zustand/blob/d27ea948843b2c955ce2be530e5e963dff8958de/docs/guides/practice-with-no-store-actions.md

**Reply by KamalakannanKrishnaraj15** [reply]
Hi, 
  I am a new user of Zustand and am attracted with the implementation. The above solution looks nice with less boilerplate compared to redux code setup.
   I need a reference for having multiple stores for better readability and maintainability approach. With logger middleware added when creating the store.
   
   @dai-shi can you help me here?

**Reply by dai-shi** [reply]
I hope someone can help. You want to open a new discussion.

**Reply by KamalakannanKrishnaraj15** [reply]
Yes

**Comment by tokland** [comment]
A class can be used:

```typescript
import { create, StoreApi } from "zustand";
import { combine } from "zustand/middleware";

type Api = StoreApi<CounterState>;

type CounterState = { value: number };

class CounterActions {
    constructor(private set: Api["setState"], private get: Api["getState"]) {}

    add = (n: number) => this.set(state => ({ value: state.value + n }));
    increment = () => this.add(+1);
    decrement = () => this.add(-1);
}

const initialState: CounterState = { value: 0 };

export const useCounterStore = create(combine(initialState, (set, get) => new CounterActions(set, get)));

// Usage

const value = useCounterStore(s => s.value);
const increment = useCounterStore(s => s.increment);
```

**Comment by devanshj** [comment]
@tokland Oooo that's very clever. Classes can be used to create circular type dependencies. In fact now you don't even need `combine`.

```ts
import create, { StoreApi as Store, Mutate } from "zustand"

class MyStore {
  counter = 0

  private add = (x: number) => {
    this.set(s => ({ counter: s.counter + x }))
  }

  increment = () => {
    // here `this` has all the actions typed (and state too ofc), exactly what the OP wanted.
    this.add(+1)
  }

  decrement = () => {
    this.add(-1)
  }


  constructor(
    private set: Store<MyStore>["setState"],
    private get: Store<MyStore>["getState"],
    private store: Mutate<Store<MyStore>, [/* add middleware mutators here */]>
  ) {}
}

let useMyStore = create<MyStore>()((set, get, store) => new MyStore(set, get, store))
```

To elaborate on "Classes can be used to create circular type dependencies"... What I mean is you can't do this...

```ts
import { StoreApi as Store } from "zustand";

const createMyStore = (set: Store<ReturnType<typeof createMyStore>>["setState"]) => {
  let counter = 0
  let add = (x: number) => set(s => ({ counter: s.counter + x }))
  let increment = () => add(1)
  let decrement = () => add(-1)

  return { counter, increment, decrement }
}

// Compilation error
// 'createMyStore' implicitly has type 'any' because it does not have a type annotation and is referenced 
//  directly or indirectly in its own initializer.
```

That is to say you can't type the parameter of a function using what it returns, but if you make that function a class, then you can do it!

```ts
import { StoreApi as Store } from "zustand";

class MyStore {
  constructor(private set: Store<MyStore>["setState"]) {}
  
  counter = 0
  private add = (x: number) => this.set(s => ({ counter: s.counter + x }))
  increment = () => this.add(1)
  decrement = () => this.add(-1)
}
```

**Reply by dai-shi** [reply]
Interesting..

**Reply by sittiponghaus** [reply]
This is dope! 😍

**Comment by PiTiLeZarD** [comment]
So essentially, a lot of legwork for something that should just work? 

```
// @ts-expect-error this obviously works
get().setBookmarks( ... ),
```
`get()` return type is `StoreProps` even if it returns `StoreProps & StoreActions` (the code works using this). Sounds like this should be fixed or I'm not understanding something fundamental (it happens very often it seems lol)

I'm creating the store like this:
```
import { create } from "zustand";
import { combine, devtools } from "zustand/middleware";

export type StoreProps = {
    /** here typedef of your variables */
};

export type StoreActionsProps = {
    /** here typedef of your actions */
};

export const InitialState: StoreProps = {
    /** here the default values for these variables */
};

const store = combine(InitialState, (set, get): StoreActionsProps => ({
    /** Implement your actions here */
}));

export const useStore = create(devtools(store));
```
Trying to let typescript do the maximum type inference as I couldn't figure out how to type set and get appropriately, but get is definitely typed wrong from within IMO 

SideNote:
```
type GetFn = () => StoreProps & StoreActionsProps;

...

(get as GetFn)().setBookmarks( ... )
```
works but you know... not ideal!

**Reply by dbritto-dev** [reply]
@PiTiLeZarD would you mind sharing us a minimal repro on stackblitz? as far I see seems you are typing wrong your store

**Reply by dai-shi** [reply]
@PiTiLeZarD
`combine` is designed to limit the usage, so its type doesn't assume calling actions from actions.
So, your type assertion should be a good workaround.
Lately, I'm working on a different pattern for more flexibility with opinions. https://github.com/zustandjs/zustand-slices
